resharper/resharper-unity/src/CSharp/Daemon/Stages/Analysis/MulOrderAnalyzer.cs

```diff
@@ -0,0 +1,124 @@
+
+
+            var (count, hasUnknownType) = CalculateMatrixIntMulCount(expression);
+            if (hasUnknownType)
+                return;
+
+            if (count > GetElementCount(expression.GetExpressionType()))
+                consumer.AddHighlighting(new InefficientMultiplyOrderWarning(expression));
+        }
+
+        private (int, bool) CalculateMatrixIntMulCount(ICSharpExpression expression)
+        {
+            if (expression == null)
+                return (0, false);
+
+            if (!(expression is IBinaryExpression binaryExpression))
+                return (0, false);
+
+            var left = binaryExpression.LeftOperand.GetOperandThroughParenthesis();
+            var right = binaryExpression.RightOperand.GetOperandThroughParenthesis();
+            if (left == null || right == null)
+                return (0, true);
+
+            var leftType = left.GetExpressionType();
+            var rightType = right.GetExpressionType();
+            if (leftType.IsUnknown || rightType.IsUnknown)
+                return (0, true);
+
+            if (!IsAcceptableType(leftType) || !IsAcceptableType(rightType))
+                return (0, true);
+
+            var leftMulCount = 0;
+            var rightMulCount = 0;
+            if (IsMatrixType(leftType))
+            {
+                var (newCount, hasUnknownType) = CalculateMatrixIntMulCount(left);
+                if (hasUnknownType)
+                    return (0, true);
+
+                leftMulCount += newCount;
+            }
+
+            if (IsMatrixType(rightType))
+            {
+                var (newCount, hasUnknownType) = CalculateMatrixIntMulCount(right);
+                if (hasUnknownType)
+                    return (0, true);
+                rightMulCount += newCount;
+            }
+
+            var myMulCount = 0;
+
+            if (IsMatrixType(leftType) && right.GetExpressionType().ToIType().IsPredefinedNumeric())
+            {
+                myMulCount += GetElementCount(leftType);
+            }
+
+            if (IsMatrixType(rightType) && left.GetExpressionType().ToIType().IsPredefinedNumeric())
+            {
+                myMulCount += GetElementCount(rightType);
+            }
+            
+            return (leftMulCount + rightMulCount + myMulCount, false);
+        }
+
+        private int GetElementCount(IExpressionType expression)
+        {
+            var name = expression.GetLongPresentableName(CSharpLanguage.Instance);
+            if (knownTypes.ContainsKey(name))
+                return knownTypes[name];
+            return 0;
+        }
+
+        private bool IsAcceptableType(IExpressionType rightType)
+        {
+            return IsMatrixType(rightType) || rightType.ToIType()?.IsPredefinedNumeric() == true;
+        }
+
+        private bool IsMatrixType([NotNull] IExpressionType expression)
+        {
+            return knownTypes.ContainsKey(expression.GetLongPresentableName(CSharpLanguage.Instance));
```

Comment:
Is `GetLongPresentableName` appropriate here? Doc comments state that it should only be used for presentation purposes.